rust/lance-encoding/src/encoder.rs

```diff
@@ -802,6 +1041,135 @@ impl FieldEncodingStrategy for CoreFieldEncodingStrategy {
     }
 }
 
+/// An encoding strategy used for 2.1+ files
+#[derive(Debug)]
+pub struct StructuralEncodingStrategy {
+    pub compression_strategy: Arc<dyn CompressionStrategy>,
+    pub version: LanceFileVersion,
+}
+
+// For some reason, clippy thinks we can add Default to the above derive but
+// rustc doesn't agree (no default for Arc<dyn Trait>)
+#[allow(clippy::derivable_impls)]
+impl Default for StructuralEncodingStrategy {
+    fn default() -> Self {
+        Self {
+            compression_strategy: Arc::<CoreArrayEncodingStrategy>::default(),
+            version: LanceFileVersion::default(),
+        }
+    }
+}
+
+impl StructuralEncodingStrategy {
+    fn is_primitive_type(data_type: &DataType) -> bool {
+        matches!(
+            data_type,
+            DataType::Boolean
+                | DataType::Date32
+                | DataType::Date64
+                | DataType::Decimal128(_, _)
+                | DataType::Decimal256(_, _)
+                | DataType::Duration(_)
+                | DataType::Float16
+                | DataType::Float32
+                | DataType::Float64
+                | DataType::Int16
+                | DataType::Int32
+                | DataType::Int64
+                | DataType::Int8
+                | DataType::Interval(_)
+                | DataType::Null
+                | DataType::Time32(_)
+                | DataType::Time64(_)
+                | DataType::Timestamp(_, _)
+                | DataType::UInt16
+                | DataType::UInt32
+                | DataType::UInt64
+                | DataType::UInt8
+                | DataType::FixedSizeBinary(_)
+                | DataType::FixedSizeList(_, _)
+                | DataType::Binary
+                | DataType::LargeBinary
+                | DataType::Utf8
+                | DataType::LargeUtf8,
+        )
+    }
+}
+
+impl FieldEncodingStrategy for StructuralEncodingStrategy {
+    fn create_field_encoder(
+        &self,
+        _encoding_strategy_root: &dyn FieldEncodingStrategy,
+        field: &Field,
+        column_index: &mut ColumnIndexSequence,
+        options: &EncodingOptions,
+    ) -> Result<Box<dyn FieldEncoder>> {
+        let data_type = field.data_type();
+        if Self::is_primitive_type(&data_type) {
+            Ok(Box::new(PrimitiveStructuralEncoder::try_new(
+                options,
+                self.compression_strategy.clone(),
+                column_index.next_column_index(field.id as u32),
+                field.clone(),
+            )?))
+        } else {
+            match data_type {
+                DataType::List(_child) | DataType::LargeList(_child) => {
+                    todo!()
+                }
```

Comment:
Will this be in a follow up PR? Or isn't hit by this code path?

Reply:
In a follow-up.  I haven't done lists / repetition levels completely yet.